packages/utils/collective/index.js

```diff
@@ -189,6 +199,241 @@ function buildNode(cliArgs, pkgJsonDict) {
+      rootTsConfig.references.push({
+        path: relative(rootPath, dirname(pkgJson._path)).replace(/\\/g, '/')
+      });
+    }
+  });
+
+  rootTsConfig.references.sort(refPathSort);
+
+  if (!existsSync(rootTsConfigPath) ||
+      !isEqual(rootTsConfig, readJsonSync(rootTsConfigPath))) {
+    writeJsonSync(
+      rootTsConfigPath,
+      sortKeys(rootTsConfig, {deep: true}),
+      {spaces: 2}
+    );
+  }
+
+  if (solo) {
+    const packagePath = dirname(Object.values(filteredPkgJsonDict)[0]._path);
+
+    const doSolo = () => {
+      const subp = spawn(typecheckBinPath, [
+        '--build',
+        '--pretty',
+        ...cliArgs
+      ], {
+        cwd: packagePath,
+        stdio: 'inherit'
+      });
+
+      subp.on('close', code => process.exit(code));
+    };
+
+    if (doClean) {
+      const subp = spawn(typecheckBinPath, [
+        '--build',
+        '--clean'
+      ], {
+        cwd: packagePath,
+        stdio: 'inherit'
+      });
+
+      subp.on('close', code => {
+        if (code) process.exit(code);
+        doSolo();
+      });
+    } else {
+      doSolo();
+    }
+  } else {
+    Object.values(filteredPkgJsonDict).forEach(pkgJson => {
+      if (pkgJson[EMBARK_COLLECTIVE] && pkgJson[EMBARK_COLLECTIVE].typecheck) {
+        collectiveTsConfig.references.push({
+          path: relative(rootPath, dirname(pkgJson._path)).replace(/\\/g, '/')
+        });
+      }
+    });
+
+    collectiveTsConfig.references.sort(refPathSort);
+
+    if (!existsSync(collectiveTsConfigPath) ||
+        !isEqual(collectiveTsConfig, readJsonSync(collectiveTsConfigPath))) {
+      writeJsonSync(
+        collectiveTsConfigPath,
+        sortKeys(collectiveTsConfig, {deep: true}),
+        {spaces: 2}
+      );
+    }
+
+    const doCollective = () => {
+      const subp = spawn(typecheckBinPath, [
+        '--build',
+        collectiveTsConfigPath,
+        '--pretty',
+        ...cliArgs
+      ], {
+        cwd: rootPath,
+        stdio: ['inherit', 'pipe', 'pipe']
+      });
+
+      subp.stdout.pipe(labeler('typecheck')).pipe(process.stdout);
+      subp.stderr.pipe(labeler('typecheck')).pipe(process.stderr);
+
+      subp.on('close', code => process.exit(code));
+    };
+
+    if (doClean) {
+      const subp = spawn(typecheckBinPath, [
+        '--build',
+        collectiveTsConfigPath,
+        '--clean'
+      ], {
+        cwd: rootPath,
+        stdio: ['inherit', 'pipe', 'pipe']
+      });
+
+      subp.stdout.pipe(labeler('typecheck')).pipe(process.stdout);
+      subp.stderr.pipe(labeler('typecheck')).pipe(process.stderr);
+
+      subp.on('close', code => {
+        if (code) process.exit(code);
+        doCollective();
+      });
+    } else {
+      doCollective();
+    }
+  }
+}
+
+function refPathSort({path: pathA}, {path: pathB}) {
+  if (pathA < pathB) {
+    return -1;
+  } else if (pathA > pathB) {
+    return 1;
+  } else {
+    return 0;
+  }
+}
+
```

Comment:
Almost everything added here is extremely hard to parse and understand, especially the body in https://github.com/embark-framework/embark/pull/2126/files#diff-b463c0f63c1daaa9c10c5fe5004b714fR226

Do you think we can make this stuff a bit more descriptive? Whether it's functions with readable names or comments or both, I'm fine either way. 

I just believe that we'll have a hard time changing/updating and making sense of this if we have to in the future.

Reply:
It is a bit hard to understand. Basically, it's meta-programming for the monorepo.

I did try to choose descriptive names for variables and functions, though maybe I can still do better. Can you give concrete examples of names that are confusing or not descriptive enough?

I'm not sure this will help, but I'll try to give a rundown:

L211 to L333 have three goals:
1. write a `tsconfig.json` into each package that's part of the collective (see first paragraph in https://github.com/embark-framework/embark/pull/2126#discussion_r357347607).
2. write a root `tsconfig.json` that is primarily useful to IDEs like vscode, i.e. in conjunction with the package-level `tsconfig.json` files.
3. if it's not a "solo run" (i.e. kicked-off by `npm run solo --typecheck` within a package and therefore scoped to that package) write a `.tsconfig.collective.json` in the root that will be used to tell `tsc` which packages to check. By default that will be all of them, but it is possible to scope from the root, e.g. `yarn typecheck --scope embark-api --scope embark-transaction-logger` or `yarn typecheck --ignore embarkjs`, etc.

The root and package-level `tsconfig.json`, but not `.tsconfig.collective.json`, are intended to be checked into version control. If structural changes are made to the monorepo (package/s renamed, added, removed) then on next run of `yarn typecheck` (or `yarn start` or `yarn watch:typecheck`) the root and relevant package-level `tsconfig.json` will be updated and those changes should be committed. To avoid unnecessary churn when `tsc` is running in watch-mode (`yarn start` or `yarn watch:typecheck`) the JSON is only written to disk if it's not a match for what's already on disk or the tsconfig file was missing.

L335 to L425 are responsible for launching the `tsc` child process.

If the `--clean` option was used then before running `tsc` to actually analyze the codease, it needs to be run with `--build --clean` and when that's finished it's run without `--clean`, since `tsc --build --clean` will only do the cleaning operations and doesn't analyze code.

The `refPathSort` function starting on L427 is used before writing the config files to ensure the ordering of the references in the configs is deterministic.